### PR TITLE
Add streaming support feature

### PR DESCRIPTION
**Pull request by ealvar3z**
Removes golangci linting deprecation warnings 
See: [Issue #49](https://github.com/sashabaranov/go-gpt3/issues/49)

### CODE REVIEW

File: completion.go [hunk 1]
```diff
@@ -106,3 +110,68 @@ func (c *Client) CreateCompletion(
...
+		return
+	}
+	defer res.Body.Close()
+
+	reader := bufio.NewReader(req.Body)
+	var line []byte
+
+	ch = make(chan string)
+
+	go func() {
```
Comment (sashabaranov) [inline completion.go:150]: There's a [recommendation](https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions) to prefer synchronous functions (also see Concurrency section [here](https://medium.com/@cep21/aspects-of-a-good-go-library-7082beabb403)).

Could we remove the goroutine from here? We can try something similar to the goroutine-free API proposal [here](https://github.com/sashabaranov/go-gpt3/issues/49#issuecomment-1373423407).

Comment (sashabaranov) [reply completion.go:150]: It might also make sense to move all streaming-related logic into separate file

Comment (ealvar3z) [reply completion.go:150]: Yes, that makes complete sense (just learned something new  lol). I will refactor per your guidelines. Thank you for your assistance.

File: stream.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	reader := bufio.NewReader(res.Body)
+	var line []byte
+	var headerData = []byte("data: ")
+
+	var responses []CompletionResponse
+	for {
```
Comment (sashabaranov) [inline stream.go:50]: This loop effectively makes it a non-streaming API. I suggest this method returns a struct like:

```go

type CompletionStream struct {
	reader *bufio.Reader
	response *http.Response
}

func (stream *CompletionStream) Recv() {
	// do the loop body logic with reader.ReadBytes, etc.
}

func (stream *CompletionStream) Close() {
	// close response body, reader, etc. (basically what you've done in defers)
}

func (c *Client) CreateCompletionStream(
	ctx context.Context,
	request CompletionRequest,
) (*CompletionStream, error) {
	// make request and construct stream, return before the loop
}
```

File: .golangci.yml [hunk 3]
```diff
@@ -149,16 +149,16 @@ linters:
   disable-all: true
   enable:
     ## enabled by default
-    - deadcode # Finds unused code
```
Comment (sashabaranov) [inline .golangci.yml:152]: Note sure it's okay to disable those

**Review by sashabaranov — CHANGES_REQUESTED** [review]

File: stream.go (hunk 2 above)
Comment (ealvar3z) [reply stream.go:50]: Very well. I will iterate and re-submit. Thank you for the guidance.

File: .golangci.yml (hunk 3 above)
Comment (ealvar3z) [reply .golangci.yml:152]: I disabled those based on the WARN messages. They give out deprecation warnings and instructs the user to remove them. I can switch them back if you prefer.

File: cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	gogpt "github.com/sashabaranov/go-gpt3"
+)
+
+func main() {
```
[the remark this answers is not in the document]
Comment (ealvar3z) [reply cmd/main.go:12]: I will remove this folder also. It was built to test the library, and I never removed it.

File: .golangci.yml (hunk 3 above)
Comment (sashabaranov) [reply .golangci.yml:152]: You are correct, however, `unused` was already enabled in the linter, so we don't need to duplicate it. Fixed in the master branch. 🙌🏻

**Comment by kkdai** [comment]
looking forward to this feature. :)

**Comment by sashabaranov** [comment]
@ealvar3z do you mind I finish this up?

**Comment by ealvar3z** [comment]
No Sir.

Thank you for your patience on this, as I haven’t been able to spend any of my off time on it.

Sent from my iPhone

On Feb 7, 2023, at 3:46 AM, sashabaranov ***@***.***> wrote:

﻿
CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


@ealvar3z<https://github.com/ealvar3z> do you mind I finish this up?

**Pull request merged**